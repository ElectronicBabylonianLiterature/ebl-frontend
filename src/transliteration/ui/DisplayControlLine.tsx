import classNames from 'classnames'
import React from 'react'
import { LineProps } from './LineProps'

export default function DisplayControlLine({
  line: { type, prefix, content },
  columns,
}: LineProps): JSX.Element {
  return (
    <>
      <td className={classNames([`Transliteration__${type}`])}>{prefix}</td>
      <td
        colSpan={columns}
        className={classNames([`Transliteration__${type}`])}
      >
        {content.map(({ value }) => value).join('')}
      </td>
    </>
  )
}
