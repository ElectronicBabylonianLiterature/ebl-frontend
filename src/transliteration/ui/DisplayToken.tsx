import React, { FunctionComponent, PropsWithChildren } from 'react'
import classNames from 'classnames'
import _ from 'lodash'
import {
  AkkadianWord,
  effectiveEnclosure,
  EgyptianMetricalFeetSeparator,
  EnclosureType,
  Gloss,
  GreekLetter,
  NamedSign,
  Sign,
  Token,
  UnknownSign,
  Variant,
  Word,
} from 'transliteration/domain/token'
import { addAccents, addBreves } from 'transliteration/domain/accents'
import { isEnclosure } from 'transliteration/domain/type-guards'
import { createModifierClasses, Modifiers } from './modifiers'
import EnclosureFlags from './EnclosureFlags'
import Flags from './Flags'
import SubIndex from 'transliteration/ui/Subindex'
import WordInfo from './WordInfo'

export type TokenWrapper = FunctionComponent<PropsWithChildren<unknown>>

interface TokenProps {
  token: Token
  Wrapper: TokenWrapper
  tokenClasses?: readonly string[]
}

function DamagedFlag({
  sign: { flags },
  Wrapper,
  children,
}: PropsWithChildren<{
  sign: { flags: readonly string[] }
  Wrapper: TokenWrapper
}>): JSX.Element {
  return flags.includes('#') ? (
    <>
      <Wrapper>
        <span className="Transliteration__bracket">⸢</span>
      </Wrapper>
      {children}
      <Wrapper>
        <span className="Transliteration__bracket">⸣</span>
      </Wrapper>
    </>
  ) : (
    <>{children}</>
  )
}

function DefaultToken({ token, Wrapper }: TokenProps): JSX.Element {
  return (
    <EnclosureFlags token={token}>
      {token.parts ? (
        token.parts.map((token, index) => (
          <DisplayToken key={index} token={token} Wrapper={Wrapper} />
        ))
      ) : isEnclosure(token) ? (
        token.value
      ) : (
        <Wrapper>{token.value}</Wrapper>
      )}
    </EnclosureFlags>
  )
}

function VariantComponent({ token, Wrapper }: TokenProps): JSX.Element {
  return (
    <>
      {(token as Variant).tokens.map((token, index) => (
        <React.Fragment key={index}>
          {index > 0 ? <Wrapper>/</Wrapper> : null}
          <DisplayToken token={token} Wrapper={Wrapper} />
        </React.Fragment>
      ))}
    </>
  )
}

function GlossComponent({ token, Wrapper }: TokenProps): JSX.Element {
  const gloss = token as Gloss
  const GlossWrapper: TokenWrapper = ({
    children,
  }: PropsWithChildren<unknown>) => (
    <Wrapper>
      <sup>{children}</sup>
    </Wrapper>
  )
  return (
    <>
      <span
        className={classNames([
          'Transliteration__glossJoiner',
          ...createModifierClasses('glossJoiner', token.enclosureType),
        ])}
      >
        <GlossWrapper>.</GlossWrapper>
      </span>
      {gloss.parts.map((token, index) =>
        isEnclosure(token) ? (
          <DisplayToken key={index} token={token} />
        ) : (
          <DisplayToken key={index} token={token} Wrapper={GlossWrapper} />
        )
      )}
    </>
  )
}

function UnknownSignComponent({ token, Wrapper }: TokenProps): JSX.Element {
  const sign = token as UnknownSign
  const signs = {
    UnclearSign: sign.enclosureType.includes('BROKEN_AWAY') ? 'o' : 'x',
    UnidentifiedSign: 'X',
  }
  return (
    <DamagedFlag sign={sign} Wrapper={Wrapper}>
      <Wrapper>
        <EnclosureFlags token={sign}>
          {signs[sign.type]}
          <Flags flags={sign.flags} />
        </EnclosureFlags>
      </Wrapper>
    </DamagedFlag>
  )
}

function EgyptianMetricalFeetSeparatorComponent({
  token,
  Wrapper,
}: TokenProps): JSX.Element {
  const sign = token as EgyptianMetricalFeetSeparator
  return (
    <DamagedFlag sign={sign} Wrapper={Wrapper}>
      <Wrapper>
        <EnclosureFlags token={token}>
          <span className="Transliteration__EgyptianMetricalFeetSeparator--colored">
            {'•'}
          </span>
          <Flags flags={sign.flags} />
        </EnclosureFlags>
      </Wrapper>
    </DamagedFlag>
  )
}

function signComponent(nameProperty: string) {
  return function SignComponent({ token, Wrapper }: TokenProps): JSX.Element {
    const sign = token as Sign
    return (
      <DamagedFlag sign={sign} Wrapper={Wrapper}>
        <Wrapper>
          <EnclosureFlags token={token}>
            {sign[nameProperty]}
            <Modifiers modifiers={sign.modifiers} />
            <Flags flags={sign.flags} />
          </EnclosureFlags>
        </Wrapper>
      </DamagedFlag>
    )
  }
}

function NamedSignComponent({ token, Wrapper }: TokenProps): JSX.Element {
  const namedSign = token as NamedSign
  const effectiveEnclosures: EnclosureType[] = effectiveEnclosure(namedSign)
  const [parts, isSubIndexConverted] = addAccents(namedSign)
  const omitSubindex = namedSign.subIndex === 1 || isSubIndexConverted
  return (
    <DamagedFlag sign={namedSign} Wrapper={Wrapper}>
      <EnclosureFlags token={namedSign} enclosures={effectiveEnclosures}>
        {parts.map((token, index) =>
          isEnclosure(token) ? (
            <DisplayToken key={index} token={token} />
          ) : (
            <DisplayToken key={index} token={token} Wrapper={Wrapper} />
          )
        )}
        <Wrapper>
          {!omitSubindex && <SubIndex token={namedSign} />}
          <Modifiers modifiers={namedSign.modifiers} />
          <Flags flags={namedSign.flags} />
        </Wrapper>
        {namedSign.sign && (
          <>
            <span className="Transliteration__bracket">(</span>
            <DisplayToken token={namedSign.sign} Wrapper={Wrapper} />
            <span className="Transliteration__bracket">)</span>
          </>
        )}
        {namedSign.surrogate && !_.isEmpty(namedSign.surrogate) && (
          <>
            <span className="Transliteration__bracket">&lt;(</span>
            {namedSign.surrogate.map((token, index) => (
              <DisplayToken key={index} token={token} Wrapper={Wrapper} />
            ))}
            <span className="Transliteration__bracket">)&gt;</span>
          </>
        )}
      </EnclosureFlags>
    </DamagedFlag>
  )
}

function GreekLetterComponent({ token, Wrapper }: TokenProps): JSX.Element {
  const letter = token as GreekLetter
  return (
    <DamagedFlag sign={letter} Wrapper={Wrapper}>
      <Wrapper>
        <EnclosureFlags token={letter}>
          {letter.letter}
          <Flags flags={letter.flags} />
        </EnclosureFlags>
      </Wrapper>
    </DamagedFlag>
  )
}

function TabulationComponent({ token, Wrapper }: TokenProps): JSX.Element {
  return (
    <Wrapper>
      <span></span>
    </Wrapper>
  )
}

function LineBreakComponent({ Wrapper }: TokenProps): JSX.Element {
  return <Wrapper>|</Wrapper>
}

function AkkadianWordComponent({
  token,
  Wrapper,
  tokenClasses: modifierClasses,
}: TokenProps): JSX.Element {
  const word = addBreves(token as AkkadianWord)
  const lastParts = _.takeRightWhile(word.parts, isEnclosure)
  const parts = _.dropRight(word.parts, lastParts.length)
  return (
    <WordInfo word={word} tokenClasses={modifierClasses ?? []}>
      <DamagedFlag sign={{ flags: word.modifiers }} Wrapper={Wrapper}>
        <EnclosureFlags token={word}>
          {parts.map((token, index) => (
            <DisplayToken key={index} token={token} Wrapper={Wrapper} />
          ))}
          <Wrapper>
            <Flags flags={word.modifiers} />
          </Wrapper>
          {lastParts.map((token, index) => (
            <DisplayToken key={index} token={token} Wrapper={Wrapper} />
          ))}
        </EnclosureFlags>
      </DamagedFlag>
    </WordInfo>
  )
}

function WordComponent({
  token,
  Wrapper,
  tokenClasses: modifierClasses,
}: TokenProps): JSX.Element {
  const word = token as Word
  return (
    <WordInfo word={word} tokenClasses={modifierClasses ?? []}>
      <EnclosureFlags token={token}>
        {word.parts.map((token, index) => (
          <DisplayToken key={index} token={token} Wrapper={Wrapper} />
        ))}
      </EnclosureFlags>
    </WordInfo>
  )
}

const tokens: ReadonlyMap<
  Token['type'],
  FunctionComponent<{
    token: Token
    Wrapper: TokenWrapper
  }>
> = new Map([
  ['UnknownNumberOfSigns', (): JSX.Element => <>…</>],
  ['Variant', VariantComponent],
  ['EgyptianMetricalFeetSeparator', EgyptianMetricalFeetSeparatorComponent],
  ['Reading', NamedSignComponent],
  ['Logogram', NamedSignComponent],
  ['Number', NamedSignComponent],
  ['Divider', signComponent('divider')],
  ['Grapheme', signComponent('name')],
  ['UnclearSign', UnknownSignComponent],
  ['UnidentifiedSign', UnknownSignComponent],
  ['Determinative', GlossComponent],
  ['PhoneticGloss', GlossComponent],
  ['LinguisticGloss', GlossComponent],
  ['Tabulation', TabulationComponent],
  ['LineBreak', LineBreakComponent],
  ['GreekLetter', GreekLetterComponent],
  ['AkkadianWord', AkkadianWordComponent],
  ['Word', WordComponent],
])

export default function DisplayToken({
  token,
  bemModifiers = [],
  Wrapper = ({ children }: PropsWithChildren<unknown>): JSX.Element => (
    <>{children}</>
  ),
}: {
  token: Token
  bemModifiers?: readonly string[]
  Wrapper?: FunctionComponent<PropsWithChildren<unknown>>
}): JSX.Element {
  const TokenComponent = tokens.get(token.type) ?? DefaultToken
  const tokenClasses = [
    `Transliteration__${token.type}`,
    ...createModifierClasses(token.type, bemModifiers),
  ]
  return (
    <span
      className={classNames([
        `Transliteration__${token.type}`,
        ...tokenClasses,
      ])}
    >
      <TokenComponent
        token={token}
        Wrapper={Wrapper}
        tokenClasses={tokenClasses}
      />
    </span>
  )
}
