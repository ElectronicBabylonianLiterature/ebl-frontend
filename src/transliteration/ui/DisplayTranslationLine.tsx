import classNames from 'classnames'
import React from 'react'
import _ from 'lodash'
import { LineProps } from './LineProps'
import TranslationLine, {
  Extent,
} from 'transliteration/domain/translation-line'
import Markup from 'transliteration/ui/markup'
import lineNumberToString from 'transliteration/domain/lineNumberToString'

function DispalyExtent({ extent }: { extent: Extent }): JSX.Element {
  const labels = extent.labels.join(' ')
  return (
    <>
      ({labels}
      {!_.isEmpty(labels) && ' '}
      {lineNumberToString(extent.number)})
    </>
  )
}

export default function DisplayTranslationLine({
  line,
  columns,
}: LineProps): JSX.Element {
  const translationLine = line as TranslationLine
  return (
    <>
      <td className={classNames([`Transliteration__${line.type}`])}>
        {translationLine.language}
        {translationLine.extent && (
          <>
            {' '}
            <DispalyExtent extent={translationLine.extent} />
          </>
        )}
        :
      </td>
      <td
        colSpan={columns}
        className={classNames([`Transliteration__${line.type}`])}
      >
        <Markup parts={translationLine.parts} />
      </td>
    </>
  )
}
