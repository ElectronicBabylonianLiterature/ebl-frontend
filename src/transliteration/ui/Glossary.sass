.Glossary
    &__lemma
        font-weight: bolder
        font-style: italic

    &__homonym
        font-weight: bolder
