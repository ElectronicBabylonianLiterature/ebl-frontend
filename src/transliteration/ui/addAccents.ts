import { NamedSign, Token, ValueToken } from 'transliteration/domain/token'

const graveAccent = '\u0300'
const acuteAccent = '\u0301'
const breve = '\u032E'
const vowels: readonly string[] = ['a', 'e', 'i', 'u', 'A', 'E', 'I', 'U']

function isValueToken(token: Token): token is ValueToken {
  return token.type === 'ValueToken'
}

function isVowel(letter: string): boolean {
  return vowels.includes(letter)
}

function addGraveAccent(letter: string): string {
  return `${letter}${graveAccent}`
}

function addAcuteAccent(letter: string): string {
  return `${letter}${acuteAccent}`
}

function addBreve(letter: string): string {
  return `${letter}${breve}`
}
class Accumulator {
  isSubIndexConverted = false
  isFirstWovel = true
  tokens: Token[] = []
  private subIndex: number | null | undefined

  constructor(subIndex: number | null | undefined) {
    this.subIndex = subIndex
  }

  addToken(token: Token): Accumulator {
    this.tokens.push(
      isValueToken(token)
        ? {
            ...token,
            value: this.addAccentsToValue(token.value),
          }
        : token
    )
    return this
  }

  private addAccentsToValue(value: string): string {
    return value
      .split('')
      .map((letter) => this.visitLetter(letter))
      .join('')
  }

  private visitLetter(letter: string): string {
    if (this.isFirstWovel && isVowel(letter)) {
      return this.visitVowel(letter)
    } else if (letter === 'h') {
      return addBreve(letter)
    } else {
      return letter
    }
  }

  private visitVowel(vowel: string): string {
    if (this.subIndex === 2) {
      this.isFirstWovel = false
      this.isSubIndexConverted = true
      return addAcuteAccent(vowel)
    } else if (this.subIndex === 3) {
      this.isFirstWovel = false
      this.isSubIndexConverted = true
      return addGraveAccent(vowel)
    } else {
      return vowel
    }
  }
}

export default function addAccents(
  namedSign: NamedSign
): readonly [readonly Token[], boolean] {
  const { tokens, isSubIndexConverted } = namedSign.nameParts.reduce(
    (acc, token) => acc.addToken(token),
    new Accumulator(namedSign.subIndex)
  )

  return [tokens, isSubIndexConverted]
}
