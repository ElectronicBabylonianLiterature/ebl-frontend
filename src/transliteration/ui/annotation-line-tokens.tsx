import React from 'react'
import {
  annotationLineAccFromColumns,
  TextLineColumn,
} from 'transliteration/domain/columns'
import { TextLine } from 'transliteration/domain/text-line'
import { LineNumber } from './line-number'

export function AnnotationLineColumns({
  line,
  lineIndex,
  columns,
  maxColumns,
}: {
  line: TextLine
  lineIndex: number
  columns: readonly TextLineColumn[]
  maxColumns: number
}): JSX.Element {
  const lineAccumulator = annotationLineAccFromColumns(columns)

  const sourceTextLine = (
    <tr className={'annotation-line__source'}>
      <td>
        <LineNumber line={line} />
      </td>
      {lineAccumulator.flatResult.map((token, index) => {
        return (
          <td key={index}>
            <span
              onClick={() =>
                console.log(
                  `clicked on token ${token.token.cleanValue} at line=${lineIndex}, index=${index}`,
                  token.token
                )
              }
            >
              {token.display()}
            </span>
          </td>
        )
      })}
    </tr>
  )
  const lemmaAnnotationLayer = (
    <tr className={'annotation-line__lemmatization'}>
      <td></td>
      {lineAccumulator.flatResult.map((token, index) => {
        return (
          <td key={index}>
            <span
              onClick={() =>
                console.log(
                  `clicked on lemma of token ${token.token.cleanValue} at line=${lineIndex}, index=${index}`,
                  token.token
                )
              }
            >
              {token.token.uniqueLemma}
            </span>
          </td>
        )
      })}
    </tr>
  )

  return (
    <>
      {sourceTextLine}
      {lemmaAnnotationLayer}
    </>
  )
}
