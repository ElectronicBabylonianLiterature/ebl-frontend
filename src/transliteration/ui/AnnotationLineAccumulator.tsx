import React from 'react'
import _ from 'lodash'
import {
  Protocol,
  Shift,
  CommentaryProtocol,
  isLeftSide,
  Token,
} from 'transliteration/domain/token'
import { isEnclosure } from 'transliteration/domain/type-guards'
import DisplayToken from './DisplayToken'
import {
  GlossWrapper,
  isCloseEnclosure,
  isOpenEnclosure,
} from './LineAccumulator'

export class MarkableToken {
  readonly token: Token
  readonly isInGloss: boolean
  readonly protocol: Protocol | null = null
  readonly language: string
  readonly hasLeadingWhitespace: boolean

  constructor(
    token: Token,
    isInGloss: boolean,
    protocol: Protocol | null,
    language: string,
    hasLeadingWhitespace?: boolean
  ) {
    this.token = token
    this.isInGloss = isInGloss
    this.protocol = protocol
    this.language = language
    this.hasLeadingWhitespace = hasLeadingWhitespace || false
  }

  display(): JSX.Element {
    return (
      <>
        {this.hasLeadingWhitespace && ' '}
        <DisplayToken
          token={this.token}
          bemModifiers={
            this.protocol === null
              ? [this.language]
              : [
                  this.language,
                  this.protocol.replace('!', 'commentary-protocol-'),
                ]
          }
          Wrapper={
            this.isInGloss && !isEnclosure(this.token)
              ? GlossWrapper
              : undefined
          }
          isInPopover={true}
        />
      </>
    )
  }
}

export interface MarkableColumnData {
  span: number | null
  content: MarkableToken[]
}

export class AnnotationLineAccumulator {
  readonly columns: MarkableColumnData[] = []
  private inGloss = false
  private language = 'AKKADIAN'
  private enclosureOpened = false
  private protocol: Protocol | null = null
  private isFirstWord = true
  lemmas: string[] = []

  get flatResult(): MarkableToken[] {
    return this.columns.flatMap((column) => column.content)
  }

  applyLanguage(token: Shift): void {
    this.language = token.language
  }

  applyCommentaryProtocol(token: CommentaryProtocol): void {
    this.protocol = token.value
  }

  pushToken(token: Token, index: number): void {
    if (_.isEmpty(this.columns)) {
      this.addColumn(1)
    }

    _.last(this.columns)?.content.push(
      new MarkableToken(
        token,
        this.inGloss,
        this.protocol,
        this.language,
        this.requireSeparator(token, index)
      )
    )
    this.enclosureOpened = isOpenEnclosure(token)
  }

  addColumn(span: number | null): void {
    this.columns.push({ span: span, content: [] })
  }

  openGloss(): void {
    this.inGloss = true
  }

  closeGloss(): void {
    this.inGloss = false
  }

  addColumnToken(token: Token, index: number): void {
    switch (token.type) {
      case 'LanguageShift':
        this.applyLanguage(token)
        break
      case 'CommentaryProtocol':
        this.applyCommentaryProtocol(token)
        break
      case 'DocumentOrientedGloss':
        isLeftSide(token) ? this.openGloss() : this.closeGloss()
        break
      case 'Column':
        throw new Error('Unexpected column token.')
      default:
        this.pushToken(token, index)
        this.isFirstWord = false
    }
  }

  private requireSeparator(token: Token, index: number): boolean {
    return (
      !this.isFirstWord && !isCloseEnclosure(token) && !this.enclosureOpened
    )
  }
}
