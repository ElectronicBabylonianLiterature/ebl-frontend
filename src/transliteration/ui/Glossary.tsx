import React from 'react'
import _ from 'lodash'
import { Link } from 'react-router-dom'
import DisplayToken from 'transliteration/ui/DisplayToken'
import { Text, GlossaryToken } from 'transliteration/domain/text'
import DictionaryWord from 'dictionary/domain/Word'
import withData from 'http/withData'
import { Promise } from 'bluebird'
import WordService from 'dictionary/application/WordService'
import produce, { castDraft } from 'immer'

function Glossary({
  data,
}: {
  data: [readonly string[], readonly GlossaryToken[]][]
}): JSX.Element {
  return (
    <section>
      <h4>Glossary</h4>
      {data.map(([lemma, tokensByLemma]) => (
        <GlossaryEntry
          key={lemma.join(' ')}
          lemma={lemma}
          tokens={tokensByLemma}
        />
      ))}
    </section>
  )
}

function GlossaryEntry({
  lemma,
  tokens,
}: {
  lemma: readonly string[]
  tokens: readonly GlossaryToken[]
}): JSX.Element {
  return (
    <div>
      <GlossaryLemma lemma={lemma} />
      <GlossaryGuideword words={tokens[0].words ?? []} />
      {': '}
      {_(tokens)
        .groupBy((token) => token.value)
        .toPairs()
        .map(([value, tokensByValue], index) => (
          <React.Fragment key={value}>
            {index > 0 && ', '}
            <GlossaryWord tokens={tokensByValue} />
          </React.Fragment>
        ))
        .value()}
    </div>
  )
}

function GlossaryLemma({ lemma }: { lemma: readonly string[] }): JSX.Element {
  return (
    <>
      {lemma.map((l, index) => (
        <span key={index}>
          {' '}
          <Link to={`/dictionary/${l}`}>{l}</Link>
        </span>
      ))}
    </>
  )
}

function GlossaryGuideword({
  words,
}: {
  words: readonly DictionaryWord[]
}): JSX.Element {
  return (
    <>
      {!_.isEmpty(words) && ', '}
      {words.map((word, index) => (
        <span key={index}>
          {index > 0 && ' '}“{word.guideWord}”
        </span>
      ))}
    </>
  )
}

function GlossaryWord({
  tokens,
}: {
  tokens: readonly GlossaryToken[]
}): JSX.Element {
  const word = _.head(tokens)?.word
  return (
    <>
      <span className="Transliteration">
        {word && <DisplayToken token={word} />}
      </span>{' '}
      ({tokens.map((token) => token.number).join(', ')})
    </>
  )
}

function isDictionaryWord(word: DictionaryWord | null): word is DictionaryWord {
  return word !== null
}

export default withData<
  {},
  { text: Text; wordService: WordService },
  [readonly string[], readonly GlossaryToken[]][]
>(Glossary, ({ text, wordService: dictionary }) => {
  return new Promise((resolve, reject) => {
    produce(text.glossary, async (draft) => {
      for (const [, tokens] of draft) {
        for (const token of tokens) {
          const words = await Promise.all(
            token.uniqueLemma.map(
              async (lemma) => await dictionary.find(lemma).catch(() => null)
            )
          )
          token.words = castDraft(words.filter(isDictionaryWord))
        }
      }
    })
      .then(resolve)
      .catch(reject)
  })
})
