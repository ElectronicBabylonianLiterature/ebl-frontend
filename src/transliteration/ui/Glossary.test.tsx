import React from 'react'
import { column, object, surface } from 'test-support/lines/at'
import { lemmatized } from 'test-support/lines/text-lemmatization'
import { render, RenderResult, screen } from '@testing-library/react'
import Glossary from './Glossary'
import WordService from 'dictionary/application/WordService'
import { Text } from 'transliteration/domain/text'
import { MemoryRouter } from 'react-router-dom'
import { createDictionaryWord } from 'test-support/glossary'
import { DictionaryContext } from 'dictionary/ui/dictionary-context'
import Bluebird from 'bluebird'

jest.mock('dictionary/application/WordService')

let element: RenderResult

beforeEach(async () => {
  const [firstLine, secondLine] = lemmatized
  const text = new Text({
    lines: [firstLine, object, surface, column, secondLine],
  })
  const wordService = new (WordService as jest.Mock<WordService>)()
  jest
    .spyOn(wordService, 'find')
    .mockImplementation(
      (id) => new Bluebird((resolve) => createDictionaryWord(id).then(resolve))
    )

  element = render(
    <MemoryRouter>
      <DictionaryContext.Provider value={wordService}>
        <Glossary text={text} wordService={wordService} />
      </DictionaryContext.Provider>
    </MemoryRouter>
  )
})

test('Glossary snapshot', async () => {
  await screen.findByText('Glossary')
  expect(element.container).toMatchSnapshot()
})
