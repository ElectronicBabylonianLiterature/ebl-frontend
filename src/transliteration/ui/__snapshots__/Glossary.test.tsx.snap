// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Glossary snapshot 1`] = `
<div>
  <section>
    <h4>
      Glossary
    </h4>
    <div>
      <a
        href="/dictionary/hepû I"
      >
        <span
          class="Glossary__lemma"
        >
          hepû
        </span>
      </a>
      , 
      “
      GW for hepû I
      ”
      : 
      <span
        class="Transliteration"
      >
        <span
          class="Transliteration__Word Transliteration__Word"
        >
          <span
            class="word-info__trigger"
            role="button"
          >
            <span
              class=""
            >
              <span
                class="Transliteration__Reading Transliteration__Reading"
              >
                <span
                  class=""
                >
                  <span
                    class="Transliteration__ValueToken Transliteration__ValueToken"
                  >
                    <span
                      class=""
                    >
                      kur
                    </span>
                  </span>
                  <sup
                    class="Transliteration__modifier"
                  />
                  <sup
                    class="Transliteration__flag"
                  />
                </span>
              </span>
            </span>
          </span>
        </span>
      </span>
       (
      l+1
      , 
      stone wig
      <sup />
       
      o
      <sup>
        ?
        !
      </sup>
       
      iii
      <sup>
        ?
      </sup>
       
      l+2
      )
    </div>
    <div>
      <a
        href="/dictionary/hepû II"
      >
        <span
          class="Glossary__lemma"
        >
          hepû
        </span>
        <span
          class="Glossary__homonym"
        >
           
          II
        </span>
      </a>
      , 
      “
      GW for hepû II
      ”
      : 
      <span
        class="Transliteration"
      >
        <span
          class="Transliteration__Word Transliteration__Word"
        >
          <span
            class="word-info__trigger"
            role="button"
          >
            <span
              class=""
            >
              <span
                class="Transliteration__Reading Transliteration__Reading"
              >
                <span
                  class=""
                >
                  <span
                    class="Transliteration__ValueToken Transliteration__ValueToken"
                  >
                    <span
                      class=""
                    >
                      kur
                    </span>
                  </span>
                  <sup
                    class="Transliteration__modifier"
                  />
                  <sup
                    class="Transliteration__flag"
                  />
                </span>
              </span>
            </span>
          </span>
        </span>
      </span>
       (
      l+1
      )
    </div>
  </section>
</div>
`;
